Tdp/how to wrap a plugin (#1)
mstergianis/docs-tap: tap-gui/configurator/external-plugins.hbs.md

to use any type of external plug-in, ensure that it is in an npm registry. This registry can be your
own private registry or a plug-in registry if you intend to use a third-party or community plug-in.
# How to Add a plugin to your Tanzu Developer Portal

In this section, we will walk you through the steps required to add a custom or community Backstage plugin to your Tanzu Developer Portal. As an example, we are using Backstage's TechInsights plugin.

The steps include:
1. Extracting the contents of the configurator image
1. Running a local verdaccio server for inner loop development
1. Creating your plugin wrapper
1. Adding your plugin wrapper to your portal configuration file
1. Issuing your portal configuration file to your Tanzu Application Platform cluster
> **Important** Tanzu Application Platform plug-ins cannot be removed from customized portals.
> However, if you decide you want to hide them, you can use the
> [runtime configuration](concepts.hbs.md#runtime) options in your `tap-values.yaml` file.

## Extracting the contents of the configurator image

Find the location of the configurator image by querying your TAP cluster

```shell

> $ kubectl -n tap-install get packages tpb.tanzu.vmware.com.0.1.2 -o jsonpath={.spec.template.spec.fetch[0].imgpkgBundle.image}

Reviewer: zsh requires quotes  for the jsonpath.
`kubectl -n tap-install get packages tpb.tanzu.vmware.com.0.1.2 -o jsonpath="{.spec.template.spec.fetch[0].imgpkgBundle.image}"`

As mentioned in [Build your Customized Tanzu Developer Portal with Configurator](building.hbs.md#prereqs),
to use any type of external plug-in, ensure that it is in an npm registry. This registry can be your

> own private registry or a plug-in registry if you intend to use a third-party or community plug-in.

Reviewer 2: @mstergianis QQ: are we enabling private registries in this release - I thought we only do public npmjs.org in 1.6.1

Author: Hmm, with these levels of workarounds a customer actually could add their own private registry by modifying the verdaccio config (without publishing directly to verdaccio) and uploading a custom builder... But yeah you're right maybe I shouldn't mention all of that.

Though FWIW I'm [reasonably confident](https://vmware.slack.com/archives/C03QP5Z3VJ4/p1689963028275329) that npmjs connectivity doesn't work currently

Reviewer 3: I was able to build with a custom plugin that I deployed to npmjs.

Author: Interesting, I was seeing issues with the verdaccio cache lifetime of `2y`. But I don't understand verdaccio all that well yet, maybe I was pointing the blame poorly. Or maybe if verdaccio is already aware of that package (metadata and maybe some cached version(s)) then it doesn't bother to download metadata. But if verdaccio has no notion of that package it will download metadata from upstream?

Anyways, more to learn for me

Reviewer 4: It worked for me as well from npmjs so it definitely should work for public npmjs published wrappers

> 

Reviewer 2: ```suggestion
# How to Add a plugin to your Tanzu Developer Portal

In this section, we will walk you through the steps required to add a custom or community Backstage plugin to your Tanzu Developer Portal. As an example, we are using Backstage's TechInsights plugin.

The steps include:
1. Extracting the contents of the builder image
1. Running a local verdaccio server for inner loop development
1. Creating your plugin wrapper
1. Adding your plugin wrapper to your portal configuration file
1. Issuing your portal configuration file to your Tanzu Application Platform cluster
```

## Extracting the contents of the builder image

Find the location of the builder image by querying your TAP cluster

```shell
$ kubectl -n tap-install get packages tpb.tanzu.vmware.com.0.1.2 -o jsonpath={.spec.template.spec.fetch[0].imgpkgBundle.image}
```

```
REGISTRY/REPOSITORY@sha256:458d9727b6f77ea2fadd2ee597885419e555e7b74057bd654f7d4eee8f006641
```

This is a carvel bundle containing the builder image, in order to get the
address of the builder image you must download the carvel bundle and look at the
image lock file within.

```shell
$ imgpkg pull -b REGISTRY/REPOSITORY@sha256:458d9727b6f77ea2fadd2ee597885419e555e7b74057bd654f7d4eee8f006641 -o bundle
$ cat bundle/.imgpkg/images.yml
```

```
---
apiVersion: imgpkg.carvel.dev/v1alpha1
images:
- annotations:
    kbld.carvel.dev/id: harbor-repo.vmware.com/esback/tdp-configurator@sha256:0b03767e0526803c055d080a7d378df9adaeaa7cff0ff5af23a2debe668f20e3
  image: REGISTRY/REPOSITORY@sha256:0b03767e0526803c055d080a7d378df9adaeaa7cff0ff5af23a2debe668f20e3
kind: ImagesLock
```

We want to download the image specified by the `image` key.

```shell

> $ imgpkg -i REGISTRY/REPOSITORY@sha256:0b03767e0526803c055d080a7d378df9adaeaa7cff0ff5af23a2debe668f20e3 -o builder

Reviewer 3: This should be `imgpkg pull -i`.  You are missing the `pull` command.

Author: Good call! Thanks @doddatpivotal 

```suggestion
$ imgpkg pull -i REGISTRY/REPOSITORY@sha256:0b03767e0526803c055d080a7d378df9adaeaa7cff0ff5af23a2debe668f20e3 -o builder
```

```

Will output the filesystem of the builder image to a directory called `builder` on your local machine.

## Running a local verdaccio server for inner loop development

Using your extracted builder from the previous step you can run a
[verdaccio](https://verdaccio.org/) server on your machine. This will enable you
to depend on typescript libraries that the Tanzu Developer Portal has developed
and are necessary for adding plugins to your Tanzu Developer Portal instance.

note: as of writing this doc a workaround is needed to adjust the parameters
of the verdaccio `offline_config.yaml`. In order to fetch recently published

> packages you can adjust the `maxage` parameter under the `npmjs` `uplinks`

Reviewer 2: We don't expect the workaround to become obsolete in TAP 1.6.1. Is this note necessary?

Author: This workaround would still be necessary in the 1.6 timeframe. Actually it's a bit of a blocker in more ways than one :cry:. This could also potentially block customers from pulling in custom plugin wrappers from NPM, since the builder image won't reach out to NPM to get metadata on new packages

section

```diff
--- old_offline_config.yaml	2023-07-21 14:14:16.182418162 -0400
+++ offline_config.yaml	2023-07-21 14:12:07.660945189 -0400
@@ -3,7 +3,7 @@
 uplinks:
   npmjs:
     url: https://registry.npmjs.org/
-    maxage: 2y
+    maxage: 2m
     cache: true
 packages:
   '@tpb/*':
```

This will enable you to fetch recent metadata and versions from npmjs.com which
will prevent errors in future steps; such as when installing lerna.

```shell
$ cd builder/builder/registry/
$ ./start_offline.sh
```

This has started a verdaccio server using the utility `forever` which runs
background processes and periodically monitors them.

If at any point you wish to deactivate the running verdaccio server you can run
```shell
$ ./node_modules/forever/bin/forever stop node_modules/verdaccio/bin/verdaccio
```

If you prefer to see the log lines verdaccio generates then you can run it
directly without forever.

```shell
$ cd builder/builder/registry/
$ ./node_modules/verdaccio/bin/verdaccio --config offline_config.yaml
```

## Creating your plugin wrappers

With your verdaccio server running, you should have access to the requisite

> libraries the Tanzu Developer Portal team developed. The only requirement is

Reviewer 2: ```suggestion
libraries of the Tanzu Developer Portal. The only requirement is
```

that you will have to tell your local machine to fetch its dependencies through
verdaccio.

> In this example we will be creating two plugin wrappers one for
> [`tech-insights`](https://github.com/backstage/backstage/tree/master/plugins/tech-insights)
> and one for
> [`tech-insights-backend`](https://github.com/backstage/backstage/tree/master/plugins/tech-insights-backend).

Reviewer 2: ```suggestion
We are going to use Backstage's TechInsights plugin as an example. This plugin is broken down into the frontend and backend, so we will be creating two plugin wrappers: one for
[`tech-insights`](https://github.com/backstage/backstage/tree/master/plugins/tech-insights)
and one for
[`tech-insights-backend`](https://github.com/backstage/backstage/tree/master/plugins/tech-insights-backend).
```

 All set! Now you might want to:
  Run the app: cd plugin-wrappers && yarn dev
  Set up the software catalog: https://backstage.io/docs/features/software-catalog/configuration
  Add authentication: https://backstage.io/docs/auth/
```

This command scaffolds a backstage project structure.

We'll start by putting into place our `.yarnrc` which will tell yarn to point to
our verdaccio server.

```shell
$ cd plugin-wrappers
$ echo 'registry "http://localhost:4873"' > .yarnrc
```

Before installing our dependencies we will remove the packages directory, which
contains a scaffolded backstage `app` and `backend`. This is because we are only
making a plugin repository and we don't need to download the dependencies of
`app` and `backend`.

```shell
$ rm -rf packages/
```

And you need to remove the packages directory from the yarn workspaces.
```diff
diff --git a/package.json b/package.json
index 00d64c9..77f38f3 100644
--- a/package.json
+++ b/package.json
@@ -24,7 +24,6 @@
   },
   "workspaces": {
     "packages": [
-      "packages/*",
       "plugins/*"
     ]
   },
```

```shell
$ yarn install
```

Which will install the backstage cli and a few other dependencies.

> ### tech-insights frontend wrapper

Reviewer 2: ```suggestion
### TechInsights frontend wrapper
```

Author: Going to style it instead as "Tech Insights" since that's how it's styled in the [backstage plugin README](https://github.com/backstage/backstage/tree/master/plugins/tech-insights#tech-insights)

```suggestion
### Tech Insights frontend wrapper
```

> Next you can use the backstage cli to create your plugin wrapper for `tech-insights`

Reviewer 2: ```suggestion
Next you can use the backstage cli to create your plugin wrapper for TechInsights
```

Author: I'd prefer not to add this suggestion. The reason I style it as `tech-insights` in this instance is because that more closely reflects the [backstage plugin name](https://github.com/backstage/backstage/blob/master/plugins/tech-insights/package.json#L2) and the [directory](https://github.com/backstage/backstage/tree/master/plugins/tech-insights) that plugin is in.

If you still think it's worthwhile to style it more like the README prose (as opposed to using a styling more similar to a technical name), then I think we go with

```suggestion
Next you can use the backstage cli to create your plugin wrapper for Tech Insights
```

import { AppPluginInterface, AppRouteSurface } from '@tpb/core-frontend';
import { SurfaceStoreInterface } from '@tpb/core-common';
import { EntityPageSurface } from '@tpb/plugin-catalog';
import React from 'react';

export const TechInsightsFrontendPlugin: AppPluginInterface =
  () => (context: SurfaceStoreInterface) => {
    context.applyWithDependency(
      AppRouteSurface,
      EntityPageSurface,
      (_appRouteSurface, entityPageSurface) => {
        entityPageSurface.servicePage.addTab(
          <EntityLayout.Route path="/techinsights" title="TechInsights">
            <EntityTechInsightsScorecardContent
              title="TechInsights Scorecard."
              description="TechInsight's default fact-checkers"
            />
          </EntityLayout.Route>,
        );
      },
    );
  };
```

And to index.ts we'll add
```ts
export { TechInsightsFrontendPlugin as plugin } from './TechInsightsFrontendPlugin';
```

The reason we alias `TechInsightsFrontendPlugin` to `plugin` is because the
Tanzu Developer Portal Configurator expects compatible plugins to export a
symbol with the name `plugin`.

```shell
$ yarn install && yarn tsc && yarn build
```

> ### tech-insights backend wrapper

Reviewer 2: ```suggestion
### TechInsights backend wrapper
```

Author: ```suggestion
### Tech Insights backend wrapper
```

that is built in, is `htpasswd`.

```yaml
packages:
  '@mycompany/*':
    access: $all
    publish: $all
```

We've added the npm organization `@mycompany/*` to our config with publish access.
This is because of the names in our package.json files we chose earlier:
`@mycompany/tech-insights-wrapper` `@mycompany/tech-insights-backend-wrapper`.

Now that authentication has been set up for your verdaccio server you can add your user with `yarn login`
```shell
$ yarn login
yarn login v1.22.19
question npm username: my-user # can be anything
question npm email: my.email@example.com # can be anything
Done in 10.06s.
```

Restart your verdaccio server and you should be able to run, from the
`plugins/tech-insights-wrapper` and the `plugins/tech-insights-backend-wrapper`
directories respectively

```
$ yarn publish
yarn publish v1.22.19
[1/4] Bumping version...
info Current version: 0.1.0
question New version: 1.0.0
[2/4] Logging in...
[3/4] Publishing...
$ backstage-cli package prepack
$ backstage-cli package postpack
success Published.
[4/4] Revoking token...
info Not revoking login token, specified via config file.
Done in 2.76s.
```

> ## Adding your plugin wrapper to your TDP configuration file

Reviewer 2: ```suggestion
## Adding your plugin wrapper to your portal configuration file
```

```yaml
app:
  plugins:
    - name: '@tpb/plugin-hello-world'
      version: '^1.6.0-release-1.6.x.1'
    - name: '@mycompany/plugin-techinsights'
      version: '1.0.0'
backend:
  plugins:
    - name: '@tpb/plugin-hello-world-backend'
      version: '^1.6.0-release-1.6.x.1'
```

> ## Issuing your TDP configuration file to your TAP cluster for building

Reviewer 2: ```suggestion
## Issuing your portal configuration file to your Tanzu Application Platform cluster for building
```
